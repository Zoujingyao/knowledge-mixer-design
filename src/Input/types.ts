import React, { ReactNode } from 'react';

export interface InputProps {
  /** 类名  */
  className?: string;
  /** 输入框的值  */
  value?: string;
  /** 输入框失去焦点的回调事件  */
  onBlur?: any;
  /** 输入框改变的回调事件  */
  onChange?: React.ChangeEventHandler<HTMLInputElement | HTMLTextAreaElement>;
  /** 提示文字  */
  placeholder?: string;
  /**
   * 允许清空
   * @default false
   */
  allowClear?: boolean;
  /**
   * 是否有边框
   * @default true
   */
  bordered?: boolean;
  /**
   * 和原生input的type一样
   *  @default text
   */
  type?: string;
  /** 输入框状态  */
  status?: null | 'error' | 'warning';
  /** 输入框大小，默认为middle  */
  size?: 'large' | 'middle' | 'small';

  disabled?: boolean;
  defaultValue?: string;
  id?: string;
  addonAfter?: ReactNode;
  addonBefore?: ReactNode;
  prefix?: ReactNode;
  suffix?: ReactNode;
}
