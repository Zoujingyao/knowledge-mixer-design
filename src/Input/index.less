@md-height: 32px;
@lg-height: 40px;
@sm-height: 24px;

.km-input {
  position: relative;
  display: inline-block;
  box-sizing: border-box;
  width: 100%;
  max-width: 300px;
  padding: 4px 11px;
  background-color: @white;
  border: 1px solid @gray500;
  border-radius: 5px;
  outline: none;
  &:focus,
  &:hover {
    border: 1px solid @primary;
    box-shadow: 0 0 2px 0 @primary;
  }
  &-border-none {
    border: none;
    &:focus,
    &:hover {
      border: none;
      box-shadow: none;
    }
  }
  // size大小
  &-md {
    // height: @md-height;
    font-size: @middle-font-size;
  }
  &-lg {
    // height: @lg-height;
    padding: 7px 11px;
    font-size: @large-font-size;
  }
  &-sm {
    // height: @sm-height;
    padding: 2px 11px;
    font-size: @small-font-size;
  }
  // 状态样式
  &-error {
    background-color: @white;
    border: 1px solid @error;
    &:focus,
    &:hover {
      border: 1px solid @error;
      box-shadow: 0 0 2px 0 @error;
    }
  }
  &-warning {
    background-color: @white;
    border: 1px solid @warning;
    &:focus,
    &:hover {
      border: 1px solid @warning;
      box-shadow: 0 0 2px 0 @warning;
    }
  }
  // 前后元素的包裹
  &-affix-wrapper {
    position: relative;
    display: inline-flex;
    // align-items: center;
    justify-content: center;
    box-sizing: border-box;
    width: 100%;
    min-width: 0;
    max-width: 300px;
    padding: 4px 11px;
    font-size: 14px;
    line-height: 1.5715;
    background-color: @white;
    background-image: none;
    border: 1px solid @gray500;
    border-radius: 5px;
    transition: all 0.3s;
    .km-input {
      padding: 2px 0;
    }
    &:focus-within,
    &:hover {
      border: 1px solid @primary;
      box-shadow: 0 0 2px 0 @primary;
    }
    &-md {
      height: 28px;
    }
    &-sm {
      height: 22px;
      input {
        font-size: 12px;
      }
    }
    &-lg {
      height: 32px;
    }
    &-error {
      border-color: @error !important;
      &:focus,
      &:hover {
        border: 1px solid @error !important;
        box-shadow: 0 0 2px 0 @error !important;
      }
    }
    &-warning {
      border-color: @warning !important;
      &:focus,
      &:hover {
        border: 1px solid @warning !important;
        box-shadow: 0 0 2px 0 @warning !important;
      }
    }
    .km-input-suffix,
    .km-input-prefix {
      display: flex;
      flex: none;
      align-items: center;
    }
    .km-input-prefix {
      margin-right: 4px;
    }
    .km-input-suffix {
      margin-left: 4px;
      .km-input-clear-icon {
        height: 100%;
        cursor: pointer;
        svg path:hover {
          fill: rgba(0, 0, 0, 0.45);
        }
        &-hidden {
          visibility: hidden;
        }
      }
    }
  }

  // 前后组件包裹
  &-wrapper {
    position: relative;
    display: table;
    box-sizing: border-box;
    width: 100%;
    margin: 0;
    padding: 0;
    font-variant: tabular-nums;
    list-style: none;
    border-collapse: separate;
    border-spacing: 0;
    .km-input-group-addon {
      position: relative;
      padding: 0 11px;
      text-align: center;
      background-color: @white;
      border: 1px solid @gray500;
      border-radius: 2px;
      transition: all 0.3s;
    }
  }
}

.km-ipt-close {
  position: absolute;
  top: 0;
  left: 0;
  z-index: 100;
  width: 20px;
  height: 20px;
}
