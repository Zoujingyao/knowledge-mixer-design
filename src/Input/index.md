---
title: Input
nav:
  title: Components
  path: /components
group:
  title: Components
  path: /components
---

# Input

## 基本样式
```tsx
import React from 'react';
import { Input } from 'knowledge-mixer-design';

export default () => {
  return(<>
    <Input size="small" placeholder="small" />
    <br />
    <br />
    <Input size="middle" placeholder="middle" />
    <br />
    <br />
    <Input size="large" placeholder="large" />
  </>);
};
```


## 不同状态的输入框
```tsx
import React from 'react';
import { Input } from 'knowledge-mixer-design';

export default () => {
return (
<>
<div>
     <Input status='error' />
     <br/>
     <br/>
     <Input status='warning' />
</div>
</>
)
};
```

## 有前置/后置的输入框
```tsx

import React from 'react';
import { Input,Check } from 'knowledge-mixer-design';

export default () =>{
// const [value,setValue] = React.useState('13123');
// const onChange = (value:any)=>{
// setValue(value);
// console.log(value);
// }
return (
 <>
 <Input prefix={<Check className='icon'/>} size='large' defaultValue="mysite" />
 <br/>
 <br/>
 <Input prefix={<Check className='icon'/>} defaultValue="mysite" />
 <br/>
 <br/>
 <Input prefix={<Check className='icon'/>} size='small' defaultValue="mysite" />
 <br/>
 <br/>
 <Input prefix={<Check className='icon'/>}  allowClear suffix={'.com'} />
 <br/>
  <br/>
   <Input prefix={<Check className='icon'/>}  status='warning' allowClear suffix={'.com'} />
 </>
 )}
 
```


<API></API>