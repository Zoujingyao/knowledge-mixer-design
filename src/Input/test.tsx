import React from 'react';
import { Input } from 'antd';

const fath = () => {
  const onChange = (val: any) => {
    console.log(val);
  };
  return <Input onChange={onChange} />;
};
