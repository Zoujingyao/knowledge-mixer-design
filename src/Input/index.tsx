import React, { FC, PropsWithChildren, useCallback, useMemo, ReactNode, useRef } from 'react';
import { InputProps } from './types';
import { getPrefixCls } from '@/utils';
import { isFunction } from 'lodash-es';
import classNames from 'classnames';
import { Close } from '@/src';
import './index.less';

const Input: FC<PropsWithChildren<InputProps>> = (props) => {
  const {
    className = '',
    id,
    onBlur,
    onChange,
    value,
    defaultValue,
    placeholder,
    allowClear = false,
    bordered = true,
    type = 'text',
    status,
    size = 'middle',
    disabled = false,
    prefix,
    suffix,
  } = props;

  const iptRef = useRef<HTMLInputElement | null>(null);

  const isAffix = allowClear || prefix !== undefined || suffix !== undefined; // 是否需要元素包裹

  const selfPrefixCls = useMemo(() => getPrefixCls('input'), []);
  const sizeCls = useMemo(() => {
    switch (size) {
      case 'middle':
        return 'md';
      case 'large':
        return 'lg';
      case 'small':
        return 'sm';
      default:
        return 'md';
    }
  }, [size]);

  // 类名汇总
  const Inputclasses = useMemo(() => {
    return classNames(
      selfPrefixCls,
      {
        [`${selfPrefixCls}-${status}`]: status && !allowClear && !prefix && !suffix,
        [`${selfPrefixCls}-${sizeCls}`]: sizeCls,
        [`${selfPrefixCls}-border-none`]: isAffix || !bordered,
      },
      className,
    );
  }, [selfPrefixCls, status, allowClear, prefix, suffix, sizeCls, isAffix, bordered, className]);
  // 两个事件的处理
  const handleOnChange = useCallback<React.ChangeEventHandler<HTMLInputElement>>(
    (e) => {
      if (isFunction(onChange)) {
        onChange(e.target.value);
      } else {
      }
    },
    [onChange],
  );
  const handleOnBlur = useCallback(
    (e: any) => {
      if (isFunction(onBlur)) onBlur(e);
    },
    [onBlur],
  );
  const textValue = value || undefined;
  // 基础Input
  const defaultInput = useMemo(
    () => (
      <input
        ref={iptRef}
        type={type}
        onChange={onChange ? handleOnChange : undefined}
        onBlur={onBlur ? handleOnBlur : undefined}
        className={Inputclasses}
        placeholder={placeholder}
        disabled={disabled}
        value={textValue}
        id={id}
        defaultValue={defaultValue || undefined}
      />
    ),
    [
      Inputclasses,
      defaultValue,
      disabled,
      handleOnBlur,
      handleOnChange,
      id,
      onBlur,
      onChange,
      placeholder,
      textValue,
      type,
    ],
  );
  // 处理清空输入框
  const handleClearClick = useCallback(() => {
    (iptRef.current as any).value = '';

    // handleOnChange(iptRef.current);
  }, [iptRef]);

  // 添加前后元素的处理
  const AffixClasses = useMemo(
    () =>
      classNames({
        [`${selfPrefixCls}-affix-wrapper-${status}`]: status,
        [`${selfPrefixCls}-affix-wrapper`]: isAffix,
        [`${selfPrefixCls}-affix-wrapper-${sizeCls}`]: sizeCls,
        [`${selfPrefixCls}-border-none`]: !bordered,
      }),
    [bordered, isAffix, selfPrefixCls, sizeCls, status],
  );

  const clearIconClassName = useMemo(
    () => classNames('km-input-clear-icon', { ['km-input-clear-icon-hidden']: value === '' }),
    [value],
  );
  const suffixSpan = useMemo(
    () =>
      !allowClear
        ? suffix && <span className="km-input-suffix">{suffix}</span>
        : suffix && (
            <span className="km-input-suffix">
              <span className={clearIconClassName}>
                {
                  <Close
                    onClick={() => {
                      handleClearClick();
                    }}
                  />
                }
              </span>
              {suffix}
            </span>
          ),
    [allowClear, suffix, clearIconClassName, handleClearClick],
  );
  const AffixInput = isAffix && (
    <span className={AffixClasses}>
      {prefix && <span className="km-input-prefix">{prefix}</span>}
      {defaultInput}
      {suffixSpan}
    </span>
  );

  return <>{isAffix ? AffixInput : defaultInput}</>;
};

export default React.memo(Input);
