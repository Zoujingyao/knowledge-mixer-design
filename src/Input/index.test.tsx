import React from 'react';
import Input from './index';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';

describe('<Input />', () => {
  it('render Input with dumi', () => {
    try {
      const ele = renderer.create(<Input size="small" placeholder="small" />).toJSON();
      expect(ele).toMatchSnapshot();
    } catch (e) {
      console.log(e);
    }
  });
});
