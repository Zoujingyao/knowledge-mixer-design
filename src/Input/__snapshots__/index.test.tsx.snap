// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> render Input with dumi 1`] = `
<input
  className="km-input km-input-sm"
  disabled={false}
  placeholder="small"
  type="text"
/>
`;
